### ISSUE TITLE
[CS] Muting players will sometimes mute multiple people.

### ISSUE THREAD

**Issue by zp** [ValveSoftware/halflife#557]
```
] version
Protocol version 48
Exe version 1.1.2.7/Stdio (cstrike)
Exe build: 10:34:27 Feb 21 2013 (5964)
```

When you mute players with the player list -> mute function, it will sometimes mute 1 or 2 more players in addition to the selected player.

Before muting: http://vp-zp.com/zp/goldsrc/not-muted.jpg
After muting: http://vp-zp.com/zp/goldsrc/muted.jpg

When I muted the highlighted player, it also muted "shulace", confirmed he was actually muted cause I was in ventrilo with him and had him attempt to talk in game.

Fairly sure this bug has been around for a while, seems to happen in servers with around 25+ players in the server.

IP of the server: 208.146.44.38:27015, it's usually full at nights. (happens on any server with around 25+ people in it though)

**Comment by 0Ky** [comment]
Yes, I've also seen this happen in Half-Life 1. When I mute players through scoreboard, I see different players being muted in a different game session. There was also a time when I unmute a player, it will unmute two players. I'm not entirely sure how to re-produce this issue.

HL version on Windows: `Exe build: 19:06:31 Oct  7 2024 (10210)`

**Comment by Splatt581** [comment]
As far as I remember, this happens because the game client mutes other players by their `hashedcdkey` field (which it receives in the `svc_updateuserinfo` message from the server). The `hashedcdkey` field is in turn the md5 hash of the `\cdkey\` key that the server receives in the `C2S_CONNECTION` message when the client connects. The client generates this key inside `CL_SendConnectPacket`, in the `CL_GetCDKeyHash` function - the client gets the md5 from its IP address.

I believe that when this mechanism was created, all clients used their own unique external addresses, but today many use NAT to access the Internet, so the client will take the md5 of `192.168.x.x` of its IP address. Local addresses under NAT can often coincide, and as a result, `\cdkey\value`, `hashedcdkey`, as well as voice mutes will coincide.

**Comment by anzz1** [comment]
I made a simple MetaMod plugin for [ReHLDS](https://github.com/rehlds/ReHLDS/) that fixes this problem: https://github.com/anzz1/revoicemutefix

It simply replaces the client-sent cdkey value hashed from local ip with a hashed value of global ip as seen from the server side, restoring the original functionality from pre-NAT era.
